src/main/java/com/sourcery/gymapp/backend/userProfile/service/UserPhotoService.java

```diff
@@ -0,0 +1,96 @@
+import org.springframework.web.multipart.MultipartFile;
+import software.amazon.awssdk.core.sync.RequestBody;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.DeleteObjectRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.UUID;
+
+@Service
+@RequiredArgsConstructor
+public class UserPhotoService {
+    private final CurrentUserService currentUserService;
+    private final UserProfileRepository userProfileRepository;
+    private final S3Client s3Client;
+
+    @Value("${aws.s3.region}")
+    private String awsRegion;
+
+    @Value("${aws.s3.bucket}")
+    private String awsBucket;
+
+    @Transactional
+    public void uploadUserPhoto(MultipartFile image) {
+        if (image.isEmpty()) {
+            throw new InvalidImageException("Image is empty", HttpStatus.BAD_REQUEST);
+        }
+        if (image.getSize() > 5242880) {
+            throw new InvalidImageException("Image is too large", HttpStatus.PAYLOAD_TOO_LARGE);
+        }
+
+        List<String> allowedContentTypes = List.of("image/jpeg", "image/png", "image/jpg", "image/gif");
+
+        if (!allowedContentTypes.contains(image.getContentType())) {
+            throw new InvalidImageException("Invalid image type", HttpStatus.UNSUPPORTED_MEDIA_TYPE);
+        }
+
+        UUID currentUserId = currentUserService.getCurrentUserId();
+
+        String bucket = awsBucket;
+        String region = awsRegion;
+
+        UserProfile userProfile = userProfileRepository.findUserProfileByUserId(currentUserId)
+                .orElseThrow(() -> new UserProfileNotFoundException(currentUserId));
+
+        if (userProfile.getAvatarUrl() != null) {
+            String objectKey = decodeObjectKeyFromAvatarUrl(userProfile.getAvatarUrl());
+            DeleteObjectRequest deleteObjectRequest = DeleteObjectRequest.builder()
+                    .bucket(bucket)
+                    .key(objectKey)
+                    .build();
+
+            s3Client.deleteObject(deleteObjectRequest);
+        }
+        String newObjectKey = currentUserId.toString() + "/user-photo-" + UUID.randomUUID();
+
+        try {
+           PutObjectRequest putObjectRequest = PutObjectRequest.builder()
+                   .bucket(bucket)
+                   .key(newObjectKey)
+                   .build();
+
+           RequestBody requestBody = RequestBody.fromInputStream(image.getInputStream(), image.getSize());
+
+           s3Client.putObject(putObjectRequest, requestBody);
+           String objectUrl = "https://%s.s3.%s.amazonaws.com/%s".formatted(bucket, region, newObjectKey);
+
+           updateAvatarUrl(userProfile, objectUrl);
+        } catch (IOException e) {
+              throw new S3PhotoUploadException();
+        }
+    }
+
+    private String decodeObjectKeyFromAvatarUrl(String avatarUrl) {
+        return avatarUrl.substring(avatarUrl.length() - 84);
```

Comment:
Same thing with this number, I suggest do something like this:
```java
 private String decodeObjectKeyFromAvatarUrl(String avatarUrl) {
        Long urlPrefixLength = 84;
        Long substringLength = avatarUrl.length() - urlPrefixLength;
        return avatarUrl.substring(substringLength);
}
```
Alternatively there are Javadoc (Tymofii used it)